rust/connlib/tunnel/src/client.rs

```diff
@@ -54,36 +60,37 @@ where
     ///
     /// Once added, when a packet for the resource is intercepted a new data channel will be created
     /// and packets will be wrapped with wireguard and sent through it.
-    pub fn add_resource(
+    pub fn add_resources(
         &mut self,
-        resource_description: ResourceDescription,
+        resources: &[ResourceDescription],
     ) -> connlib_shared::Result<()> {
-        if let Some(resource) = self.role_state.resource_ids.get(&resource_description.id()) {
-            if resource.has_different_address(resource) {
-                self.remove_resource(resource.id());
+        for resource_description in resources {
+            if let Some(resource) = self.role_state.resource_ids.get(&resource_description.id()) {
+                if resource.has_different_address(resource) {
+                    self.remove_resource(resource.id());
+                }
             }
-        }
 
-        match &resource_description {
-            ResourceDescription::Dns(dns) => {
-                self.role_state
-                    .dns_resources
-                    .insert(dns.address.clone(), dns.clone());
+            match &resource_description {
+                ResourceDescription::Dns(dns) => {
+                    self.role_state
+                        .dns_resources
+                        .insert(dns.address.clone(), dns.clone());
+                }
+                ResourceDescription::Cidr(cidr) => {
+                    self.role_state
+                        .cidr_resources
+                        .insert(cidr.address, cidr.clone());
+                }
             }
-            ResourceDescription::Cidr(cidr) => {
-                self.add_route(cidr.address)?;
 
-                self.role_state
-                    .cidr_resources
-                    .insert(cidr.address, cidr.clone());
-            }
+            self.role_state
+                .resource_ids
+                .insert(resource_description.id(), resource_description.clone());
         }
 
-        self.role_state
-            .resource_ids
-            .insert(resource_description.id(), resource_description);
-
         self.update_resource_list()?;
+        self.update_routes()?;
```

Comment:
So the only thing in this function that can fail is calling the callbacks? Can these callbacks actually fail and is there something meaningful we can do with the error? Why not log within the callback but continue?

Reply:
For `update_routes` also netlink could fail or the win32 api call.

But still nothing meaningful we can do with the error.

Reply:
But this will require some changes in each platform I will do it on a follow up in #4107

Reply:
Though, thinking more about it, probably in the future we want to disconnect/exit if `on_update_routes` fails. If that fails we don't have any routes and we can't work properly.

Reply:
Same goes for `get_system_default_resolvers`